[omitempty] for strings causes compilation error

<!-- It is advisable to update all relevant modules using `v outdated` and `v install` -->
**V doctor:**
```
OS: linux, Pop!_OS 21.04
Processor: 8 cpus, 64bit, little endian, Intel(R) Core(TM) i7-8550U CPU @ 1.80GHz
CC version: cc (Ubuntu 10.3.0-1ubuntu1) 10.3.0

getwd: /home/me/Desktop
vmodules: /home/me/.vmodules
vroot: /home/me/REPOS/v/v
vexe: /home/me/REPOS/v/v/v
vexe mtime: 2021-11-24 19:19:36
is vroot writable: true
is vmodules writable: true
V full version: V 0.2.4 fbe2b5c.11d7062

Git version: git version 2.30.2
Git vroot status: weekly.2021.47-15-g11d70624
.git/config present: true
thirdparty/tcc status: thirdparty-linux-amd64 333c14de

```

**What did you do?**
`v -g -o vdbg cmd/v && vdbg omitempty.v`
```v
import json

struct Foo {
	foo int [omitempty]
}

struct Bar {
	foo string [omitempty] // This line causes the file not to compile
}

five := json.encode(&Foo{5})
println('$five') // {"foo":5}

empty := json.encode(&Foo{0})
println('$empty') // {}

empty_str := json.encode(&Bar{})
println('$empty_str') // EXPECTED: {}

```

**What did you expect to see?**

The file to compile


**What did you see instead?**
```
==================
/tmp/v_1000/omitempty.15370911601244302045.tmp.c:1983: error: invalid operand types for binary operation
/tmp/v_1000/omitempty.15370911601244302045.tmp.c:1983: error: invalid aggregate type for register load
...
==================
(Use `v -cg` to print the entire error message)

builder error: 
==================
C error. This should never happen.

This is a compiler bug, please report it using `v bug file.v`.

https://github.com/vlang/v/issues/new/choose

You can also use #help on Discord: https://discord.gg/vlang

```

--- a/vlib/json/json_omitempty_test.v
+++ b/vlib/json/json_omitempty_test.v
@@ -1,0 +1,18 @@
+import json
+
+pub struct MyStruct {
+pub mut:
+	code    int
+	message string
+	data    string [omitempty]
+	// data2 ?string [omitempty]
+}
+
+fn test_simple() {
+	obj := MyStruct{
+		code: 1
+		message: 'yes'
+		// data2: "a"
+	}
+	dump(json.encode(obj))
+}

--- a/vlib/json/json_omitempty_test.v
+++ b/vlib/json/json_omitempty_test.v
@@ -4,15 +4,22 @@ pub struct MyStruct {
 pub mut:
 	code    int
 	message string
-	data    string [omitempty]
-	// data2 ?string [omitempty]
+	data    string  [omitempty]
+	data2   ?string [omitempty]
 }
 
 fn test_simple() {
 	obj := MyStruct{
 		code: 1
 		message: 'yes'
-		// data2: "a"
+		data2: 'a'
 	}
-	dump(json.encode(obj))
+	assert dump(json.encode(obj)) == '{"code":1,"message":"yes","data2":"a"}'
+}
+
+fn test_none() {
+	obj := MyStruct{
+		code: 1
+	}
+	assert dump(json.encode(obj)) == '{"code":1,"message":""}'
 }

--- a/vlib/v/gen/c/json.v
+++ b/vlib/v/gen/c/json.v
@@ -647,11 +647,18 @@ fn (mut g Gen) gen_struct_enc_dec(utyp ast.Type, type_info ast.TypeInfo, styp st
 			'val'
 		}
 		is_option := field.typ.has_flag(.option)
+		indent := if is_option { '\t\t' } else { '\t' }
 		if is_option {
 			enc.writeln('\tif (val${op}${c_name(field.name)}.state != 2) {')
 		}
 		if is_omit_empty {
-			enc.writeln('\t if (val${op}${c_name(field.name)} != ${g.type_default(field.typ)})')
+			if field.typ.has_flag(.option) {
+				enc.writeln('${indent}if (val${op}${c_name(field.name)}.state != 2)')
+			} else if field.typ == ast.string_type {
+				enc.writeln('${indent}if (val${op}${c_name(field.name)}.len != 0)')
+			} else {
+				enc.writeln('${indent}if (val${op}${c_name(field.name)} != ${g.type_default(field.typ)})')
+			}
 		}
 		if !is_js_prim(field_type) {
 			if field_sym.kind == .alias {
@@ -662,44 +669,44 @@ fn (mut g Gen) gen_struct_enc_dec(utyp ast.Type, type_info ast.TypeInfo, styp st
 		if field_sym.kind == .enum_ {
 			if g.is_enum_as_int(field_sym) {
 				if field.typ.has_flag(.option) {
-					enc.writeln('\tcJSON_AddItemToObject(o, "${name}", json__encode_u64(*${prefix_enc}${op}${c_name(field.name)}.data));\n')
+					enc.writeln('${indent}\tcJSON_AddItemToObject(o, "${name}", json__encode_u64(*${prefix_enc}${op}${c_name(field.name)}.data));\n')
 				} else {
-					enc.writeln('\tcJSON_AddItemToObject(o, "${name}", json__encode_u64(${prefix_enc}${op}${c_name(field.name)}));\n')
+					enc.writeln('${indent}\tcJSON_AddItemToObject(o, "${name}", json__encode_u64(${prefix_enc}${op}${c_name(field.name)}));\n')
 				}
 			} else {
 				if field.typ.has_flag(.option) {
-					enc.writeln('\t{')
-					enc.writeln('\t\tcJSON *enum_val;')
+					enc.writeln('${indent}\t{')
+					enc.writeln('${indent}\t\tcJSON *enum_val;')
 					g.gen_enum_to_str(field.typ, field_sym, '*(${g.base_type(field.typ)}*)${prefix_enc}${op}${c_name(field.name)}.data',
-						'enum_val', '\t\t', mut enc)
-					enc.writeln('\t\tcJSON_AddItemToObject(o, "${name}", enum_val);')
-					enc.writeln('\t}')
+						'enum_val', '${indent}\t\t', mut enc)
+					enc.writeln('${indent}\t\tcJSON_AddItemToObject(o, "${name}", enum_val);')
+					enc.writeln('${indent}\t}')
 				} else {
-					enc.writeln('\t{')
-					enc.writeln('\t\tcJSON *enum_val;')
+					enc.writeln('${indent}\t{')
+					enc.writeln('${indent}\t\tcJSON *enum_val;')
 					g.gen_enum_to_str(field.typ, field_sym, '${prefix_enc}${op}${c_name(field.name)}',
-						'enum_val', '\t\t', mut enc)
-					enc.writeln('\t\tcJSON_AddItemToObject(o, "${name}", enum_val);')
-					enc.writeln('\t}')
+						'enum_val', '${indent}\t\t', mut enc)
+					enc.writeln('${indent}\t\tcJSON_AddItemToObject(o, "${name}", enum_val);')
+					enc.writeln('${indent}\t}')
 				}
 			}
 		} else {
 			if field_sym.name == 'time.Time' {
 				// time struct requires special treatment
 				// it has to be encoded as a unix timestamp number
-				enc.writeln('\tcJSON_AddItemToObject(o, "${name}", json__encode_u64(${prefix_enc}${op}${c_name(field.name)}._v_unix));')
+				enc.writeln('${indent}\tcJSON_AddItemToObject(o, "${name}", json__encode_u64(${prefix_enc}${op}${c_name(field.name)}._v_unix));')
 			} else {
 				if !field.typ.is_real_pointer() {
-					enc.writeln('\tcJSON_AddItemToObject(o, "${name}", ${enc_name}(${prefix_enc}${op}${c_name(field.name)})); /*A*/')
+					enc.writeln('${indent}\tcJSON_AddItemToObject(o, "${name}", ${enc_name}(${prefix_enc}${op}${c_name(field.name)})); /*A*/')
 				} else {
 					arg_prefix := if field.typ.is_ptr() { '' } else { '*' }
 					sptr_value := '${prefix_enc}${op}${c_name(field.name)}'
 					if !field.typ.has_flag(.option) {
-						enc.writeln('\tif (${sptr_value} != 0) {')
-						enc.writeln('\t\tcJSON_AddItemToObject(o, "${name}", ${enc_name}(${arg_prefix}${sptr_value}));')
-						enc.writeln('\t}\n')
+						enc.writeln('${indent}\tif (${sptr_value} != 0) {')
+						enc.writeln('${indent}\t\tcJSON_AddItemToObject(o, "${name}", ${enc_name}(${arg_prefix}${sptr_value}));')
+						enc.writeln('${indent}\t}\n')
 					} else {
-						enc.writeln('\t\tcJSON_AddItemToObject(o, "${name}", ${enc_name}(${arg_prefix}${sptr_value}));')
+						enc.writeln('${indent}\t\tcJSON_AddItemToObject(o, "${name}", ${enc_name}(${arg_prefix}${sptr_value}));')
 					}
 				}
 			}

